servicetalk-loadbalancer/src/main/java/io/servicetalk/loadbalancer/RoundRobinLoadBalancer.java

```diff
@@ -202,6 +223,42 @@ public void onNext(final ServiceDiscovererEvent<ResolvedAddress> event) {
                 }
             }
 
+            @SuppressWarnings("unchecked")
+            private Host<ResolvedAddress, C> createHost(ResolvedAddress addr) {
+                Host<ResolvedAddress, C> host = new Host<>(addr);
+                host.onClose().afterFinally(() ->
+                        activeHostsUpdater.updateAndGet(RoundRobinLoadBalancer.this,
+                                previousHosts -> removeFromHostsList(previousHosts, addr)
+                        )
+                ).subscribe();
+                return host;
+            }
+
+            private List<Host<ResolvedAddress, C>> removeFromHostsList(
+                    List<Host<ResolvedAddress, C>> oldHosts, ResolvedAddress addr) {
+                @SuppressWarnings("unchecked")
+                final List<Host<ResolvedAddress, C>> oldHostsTyped = (List<Host<ResolvedAddress, C>>) oldHosts;
+                if (oldHostsTyped.isEmpty()) {
+                    // this can happen when an expired host is removed,
+                    // but all of its connections have already been closed
```

Comment:
That is true. The code here could be used to reflect a future scenario where `expired` events can be emitted from SD. Followed by an unavailable event which would actually mean removal, which in such case would cause entering this check. For now, this is just a double check. I will update the comment as recommended.